igvm/vm.py

```diff
@@ -588,29 +605,26 @@ def aws_build(self,
 
     def rename(self, new_hostname):
         """Rename the VM"""
-        self.dataset_obj['hostname'] = new_hostname
-        self.check_serveradmin_config()
-
-        fd = BytesIO()
-        fd.write(bytes(new_hostname, 'utf-8'))
-        self.put('/etc/hostname', fd)
-        self.put('/etc/mailname', fd)
-
-        hosts_file = [
-            line
-            for line in self.run('cat /etc/hosts').splitlines()
-            if not line.startswith(str(self.dataset_obj['intern_ip']))
-        ]
-        hosts_file.append('{0}\t{1}'.format(
-            self.dataset_obj['intern_ip'], new_hostname
-        ))
-        self.run("echo '{0}' > /etc/hosts".format('\n'.join(hosts_file)))
-
         with Transaction() as transaction:
+            self.set_hostname(new_hostname, transaction=transaction)
+            self.check_serveradmin_config()
+
             self.shutdown(transaction=transaction)
+
             self.hypervisor.redefine_vm(self, new_fqdn=new_hostname)
+            transaction.on_rollback(
+                'undo_redefine',
+                self.hypervisor.redefine_vm,
+                self,
+                new_fqdn=self.previous_hostname
+            )
 
-            self.dataset_obj.commit()
+            self.hypervisor.mount_vm_storage(self, transaction=transaction)
+
+            time.sleep(5)
```

Comment:
what do we need this `sleep` for now?

Reply:
Serveradmin takes some time to commit the data. 

If we start the puppetrun straight away, the Puppet ENC will not see the new hostname there.

Reply:
but as far as i understand the `commit` call is blocking. is this really a problem?

Reply:
I'll retest, but afair, yesterday it caused me problems.

Reply:
@Woellchen is right. Committing to Serveradmin with adminapi is a blocking operation. There is no way the function returns before the data is safely committed to the database unless you run it in a synchronous way manually.